### ISSUE TITLE
optimize.minimize: new argument that calculates function value and derivative together

### ISSUE THREAD

**Issue by dlwocks** [scipy/scipy#6904]
Some context first:
I was implementing feedforward neural network with backpropagation. To train network, I used `optimize.minimize` (`method='BFGS'` in particular) to find the weight(`x`) that minimizes cost function for given input and target answer (input and target are constant throughout minimization). The weight that is used to compute cost function(`fun`) and its derivative(`jac`) constantly changes inside the `optimize.minimze`; but, repetitively, `fun(x)` and `jac(x)` is calculated separately for same value of `x`. Both functions require the result from forward propagation(which only depends on value of `x`), so calculating them separately created unnecessary computation. 
So, my main point:
I though of inputting new argument(which is a function) that calculates `fun` and `jac` together instead of calculating them separately. It will be a function that returns `(fun(x), jac(x))`, but internally minimizes repetitive computation.
Some concerns:
I don't understand the exact logic of minimize algorithm, so I'm not sure whether this idea fits into the logic or not. I mean, are there algorithms or situation where no same `x` is used to calculate both `fun` and `jac`, or where `fun` and `jac` _have_ to be calculated separately? I hope someone understands the logic can tell me whether this idea make sense or not.

Also, I'm new to GitHub community, I'm not expert in python or optimization problem, I'm not a native English speaker, so please tell me if I got anything basic wrong. Thanks!

**Comment by pv** [comment]
Using a function that computes both function and gradient at the same
time is already possible, set `jac=True`.
